## 探索电影数据集

在这个项目中，你将尝试使用所学的知识，使用 `NumPy`、`Pandas`、`matplotlib`、`seaborn` 库中的函数，来对电影数据集进行探索。

下载数据集：
[TMDb电影数据](https://s3.cn-north-1.amazonaws.com.cn/static-documents/nd101/explore+dataset/tmdb-movies.csv)



数据集各列名称的含义：
<table>
<thead><tr><th>列名称</th><th>id</th><th>imdb_id</th><th>popularity</th><th>budget</th><th>revenue</th><th>original_title</th><th>cast</th><th>homepage</th><th>director</th><th>tagline</th><th>keywords</th><th>overview</th><th>runtime</th><th>genres</th><th>production_companies</th><th>release_date</th><th>vote_count</th><th>vote_average</th><th>release_year</th><th>budget_adj</th><th>revenue_adj</th></tr></thead><tbody>
 <tr><td>含义</td><td>编号</td><td>IMDB 编号</td><td>知名度</td><td>预算</td><td>票房</td><td>名称</td><td>主演</td><td>网站</td><td>导演</td><td>宣传词</td><td>关键词</td><td>简介</td><td>时常</td><td>类别</td><td>发行公司</td><td>发行日期</td><td>投票总数</td><td>投票均值</td><td>发行年份</td><td>预算（调整后）</td><td>票房（调整后）</td></tr>
</tbody></table>


**请注意，你需要提交该报告导出的 `.html`、`.ipynb` 以及 `.py` 文件。**



---

---

## 第一节 数据的导入与处理

在这一部分，你需要编写代码，使用 Pandas 读取数据，并进行预处理。


**任务1.1：** 导入库以及数据

1. 载入需要的库 `NumPy`、`Pandas`、`matplotlib`、`seaborn`。
2. 利用 `Pandas` 库，读取 `tmdb-movies.csv` 中的数据，保存为 `movie_data`。

提示：记得使用 notebook 中的魔法指令 `%matplotlib inline`，否则会导致你接下来无法打印出图像。

In [1]:
import numpy as np
import pandas as pd
import matplotlib
import seaborn as sb

movie_data = pd.read_csv("tmdb-movies.csv")
%matplotlib inline

---

**任务1.2: ** 了解数据

你会接触到各种各样的数据表，因此在读取之后，我们有必要通过一些简单的方法，来了解我们数据表是什么样子的。

1. 获取数据表的行列，并打印。
2. 使用 `.head()`、`.tail()`、`.sample()` 方法，观察、了解数据表的情况。
3. 使用 `.dtypes` 属性，来查看各列数据的数据类型。
4. 使用 `isnull()` 配合 `.any()` 等方法，来查看各列是否存在空值。
5. 使用 `.describe()` 方法，看看数据表中数值型的数据是怎么分布的。



In [2]:
for item in movie_data.columns:
    print(item, ':', movie_data[item].dtype)
print(movie_data.all())
print(movie_data.isnull().any())
print(movie_data.describe())

id : int64
imdb_id : object
popularity : float64
budget : int64
revenue : int64
original_title : object
cast : object
homepage : object
director : object
tagline : object
keywords : object
overview : object
runtime : int64
genres : object
production_companies : object
release_date : object
vote_count : int64
vote_average : float64
release_year : int64
budget_adj : float64
revenue_adj : float64
id                       True
imdb_id                  True
popularity               True
budget                  False
revenue                 False
original_title           True
cast                     True
homepage                 True
director                 True
tagline                  True
keywords                 True
overview                 True
runtime                 False
genres                   True
production_companies     True
release_date             True
vote_count               True
vote_average             True
release_year             True
budget_adj              False
rev

---

**任务1.3: ** 清理数据

在真实的工作场景中，数据处理往往是最为费时费力的环节。但是幸运的是，我们提供给大家的 tmdb 数据集非常的「干净」，不需要大家做特别多的数据清洗以及处理工作。在这一步中，你的核心的工作主要是对数据表中的空值进行处理。你可以使用 `.fillna()` 来填补空值，当然也可以使用 `.dropna()` 来丢弃数据表中包含空值的某些行或者列。

任务：使用适当的方法来清理空值，并将得到的数据保存。

In [3]:
movie_data.dropna(inplace = True)
movie_data.to_csv("tmdb-movies-adjusted.csv")

---

---

## 第二节 根据指定要求读取数据


相比 Excel 等数据分析软件，Pandas 的一大特长在于，能够轻松地基于复杂的逻辑选择合适的数据。因此，如何根据指定的要求，从数据表当获取适当的数据，是使用 Pandas 中非常重要的技能，也是本节重点考察大家的内容。



---

**任务2.1: ** 简单读取

1. 读取数据表中名为 `id`、`popularity`、`budget`、`runtime`、`vote_average` 列的数据。
2. 读取数据表中前1～20行以及48、49行的数据。
3. 读取数据表中第50～60行的 `popularity` 那一列的数据。

要求：每一个语句只能用一行代码实现。

In [4]:
movie_data[['id', 'budget', 'runtime', 'vote_average']]

movie_data.iloc[[x for x in range(1, 50) if x<=20 or x>=48]]

movie_data['popularity'].iloc[50:60]

64    2.395366
65    2.354654
66    2.345821
67    2.331636
68    2.331567
71    2.281350
75    2.141506
76    2.140199
77    2.121884
78    2.096273
Name: popularity, dtype: float64

---

**任务2.2: **逻辑读取（Logical Indexing）

1. 读取数据表中 **`popularity` 大于5** 的所有数据。
2. 读取数据表中 **`popularity` 大于5** 的所有数据且**发行年份在1996年之后**的所有数据。

提示：Pandas 中的逻辑运算符如 `&`、`|`，分别代表`且`以及`或`。

要求：请使用 Logical Indexing实现。

In [5]:
movie_data[movie_data['popularity'] > 5]

movie_data[(movie_data['popularity'] > 5) & (movie_data['release_year'] > 1996)]

id    imdb_id  popularity     budget     revenue  \
0     135397  tt0369610   32.985763  150000000  1513528810   
1      76341  tt1392190   28.419936  150000000   378436354   
2     262500  tt2908446   13.112507  110000000   295238201   
3     140607  tt2488496   11.173104  200000000  2068178225   
4     168259  tt2820852    9.335014  190000000  1506249360   
5     281957  tt1663202    9.110700  135000000   532950503   
6      87101  tt1340138    8.654359  155000000   440603537   
7     286217  tt3659388    7.667400  108000000   595380321   
8     211672  tt2293640    7.404165   74000000  1156730962   
9     150540  tt2096673    6.326804  175000000   853708609   
10    206647  tt2379713    6.200282  245000000   880674609   
11     76757  tt1617661    6.189369  176000003   183987723   
12    264660  tt0470752    6.118847   15000000    36869414   
13    257344  tt2120120    5.984995   88000000   243637091   
14     99861  tt2395427    5.944927  280000000  1405035767   
15    273248  tt3460252    5.898400   44000000   155760117   
16    260346  tt2446042    5.749758   48000000   325771424   
17    102899  tt0478970    5.573184  130000000   518602163   
19    131634  tt1951266    5.476958  160000000   650523427   
20    158852  tt1964418    5.462138  190000000   209035668   
629   157336  tt0816692   24.949134  165000000   621752480   
630   118340  tt2015381   14.311205  170000000   773312399   
631   100402  tt1843866   12.971027  170000000   714766572   
632   245891  tt2911666   11.422751   20000000    78739897   
633   131631  tt1951265   10.739009  125000000   752100229   
634   122917  tt2310332   10.174599  250000000   955119788   
635   177572  tt2245084    8.691294  165000000   652105443   
636   205596  tt2084970    8.110711   14000000   233555708   
637   198663  tt1790864    7.137273   34000000   348319861   
639   210577  tt2267998    6.438727   61000000   369330363   
...      ...        ...         ...        ...         ...   
1387   16869  tt0361748    6.098027   70000000   319131050   
1389     767  tt0417741    5.076472  250000000   933959197   
1919   27205  tt1375666    9.363643  160000000   825500000   
1920   10138  tt1228705    5.704860  200000000   623933331   
1921   12155  tt1014759    5.572950  200000000  1025467110   
1922   44214  tt0947798    5.293180   13000000   327803731   
2409     550  tt0137523    8.947905   63000000   100853753   
2410     603  tt0133093    7.753899   63000000   463517383   
2633     120  tt0120737    8.575419   93000000   871368364   
2634     671  tt0241527    8.021423  125000000   976475550   
2875     155  tt0468569    8.466668  185000000  1001921825   
2876   10681  tt0910970    5.678119  180000000   521311860   
3372    1771  tt0458339    7.959228  140000000   370569774   
3374   12445  tt1201607    5.711315  125000000  1327817822   
3911     121  tt0167261    8.095275   79000000   926287400   
4361   24428  tt0848228    7.637767  220000000  1519557910   
4362   52520  tt1496025    7.031452   70000000   132400000   
4363   49026  tt1345836    6.591277  250000000  1081041287   
4364   68718  tt1853728    5.944518  100000000   425368238   
4365   37724  tt1074638    5.603587  200000000  1108561013   
4949     122  tt0167260    7.122455   94000000  1118888979   
4950     277  tt0320691    6.887883   22000000    95708457   
4951      22  tt0325980    6.864067  140000000   655011224   
4952      24  tt0266697    6.174132   30000000   180949000   
5422  109445  tt2294629    6.112766  150000000  1274219009   
5423   49047  tt1454468    5.242753  105000000   716392705   
5424   76338  tt1981115    5.111900  170000000   479765000   
6190     674  tt0330373    5.939927  150000000   895921036   
6191     272  tt0372784    5.400826  150000000   374218673   
6554     834  tt0401855    5.838503   50000000   111340801   

                                         original_title  \
0                                        Jurassic World   
1                                    Mad Max: F

---

**任务2.3: **分组读取

1. 对 `release_year` 进行分组，使用 [`.agg`](http://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.core.groupby.DataFrameGroupBy.agg.html) 获得 `revenue` 的均值。
2. 对 `director` 进行分组，使用 [`.agg`](http://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.core.groupby.DataFrameGroupBy.agg.html) 获得 `popularity` 的均值，从高到低排列。

要求：使用 `Groupby` 命令实现。

In [15]:
avg_revenue = movie_data.groupby(['release_year']).agg(np.mean)['revenue']
print (avg_revenue)
avg_popularity = movie_data.groupby(['director']).agg(np.mean)['popularity'].sort_values(ascending = False)
print (avg_popularity)

release_year
1961    4.000000e+06
1962    5.960000e+07
1963    7.889876e+07
1964    6.295027e+07
1965    1.411957e+08
1967    1.115848e+08
1969    8.197449e+07
1970    0.000000e+00
1971    4.588650e+07
1972    2.450664e+08
1973    3.015420e+08
1974    4.878600e+07
1975    1.961918e+08
1976    0.000000e+00
1977    3.202789e+08
1978    7.233049e+07
1979    9.248203e+07
1980    2.798892e+08
1981    1.219442e+08
1982    2.798207e+08
1983    1.699437e+08
1984    1.470656e+08
1985    1.413097e+08
1986    5.500000e+06
1987    5.718027e+07
1988    9.451583e+06
1989    2.005648e+08
1990    1.429964e+08
1991    1.282618e+08
1992    2.354249e+08
1993    1.497505e+08
1994    1.363249e+08
1995    1.216781e+08
1996    1.127452e+08
1997    3.008943e+08
1998    1.034330e+08
1999    2.073004e+08
2000    1.043420e+08
2001    2.270173e+08
2002    1.788673e+08
2003    1.368929e+08
2004    1.738350e+08
2005    8.538829e+07
2006    1.021119e+08
2007    1.133194e+08
2008    8.715175e+07
2009    9.903001e+07


---

---

## 第三节 绘图与可视化

接着你要尝试对你的数据进行图像的绘制以及可视化。这一节最重要的是，你能够选择合适的图像，对特定的可视化目标进行可视化。所谓可视化的目标，是你希望从可视化的过程中，观察到怎样的信息以及变化。例如，观察票房随着时间的变化、哪个导演最受欢迎等。

<table>
<thead><tr><th>可视化的目标</th><th>可以使用的图像</th></tr></thead><tbody>
 <tr><td>表示某一属性数据的分布</td><td>饼图、直方图、散点图</td></tr>
 <tr><td>表示某一属性数据随着某一个变量变化</td><td>条形图、折线图、热力图</td></tr>
 <tr><td>比较多个属性的数据之间的关系</td><td>散点图、小提琴图、堆积条形图、堆积折线图</td></tr>
</tbody></table>

在这个部分，你需要根据题目中问题，选择适当的可视化图像进行绘制，并进行相应的分析。对于选做题，他们具有一定的难度，你可以尝试挑战一下～

**任务3.1：**对 `popularity` 最高的20名电影绘制其 `popularity` 值。

---
**任务3.2：**分析电影净利润（票房-成本）随着年份变化的情况，并简单进行分析。

---

**[选做]任务3.3：**选择最多产的10位导演（电影数量最多的），绘制他们排行前3的三部电影的票房情况，并简要进行分析。

---

**[选做]任务3.4：**分析1968年~2015年六月电影的数量的变化。

---

**[选做]任务3.5：**分析1968年~2015年六月电影 `Comedy` 和 `Drama` 两类电影的数量的变化。

> 注意: 当你写完了所有的代码，并且回答了所有的问题。你就可以把你的 iPython Notebook 导出成 HTML 文件。你可以在菜单栏，这样导出**File -> Download as -> HTML (.html)、Python (.py)** 把导出的 HTML、python文件 和这个 iPython notebook 一起提交给审阅者。